### PR TITLE
HIP-584: Implement `AssociatePrecompile`

### LINKED ISSUE #5648: HIP-584: Provide implementation for associate and dissociate precompiles
### Problem

We need default implementations for precompiles related to associate and dissociate logic, so we should copy needed logic from hedera-services into hedera-mirror-web3.

### Solution

Copy precompile logic for the following precompiles:

- AbstractAssociatePrecompile
- AssociatePrecompile
- AbstractDissociatePrecompile
- DissociatePrecompile

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by IvanKavaldzhiev**
**Description**:

Implementing `AssociatePrecompile` using `AsssociateLogic` from another PR.

**Related issue(s)**:

Fixes #5648

**Notes for reviewer**:
Needs the changes from this [PR](https://github.com/hashgraph/hedera-mirror-node/pull/6032) to work properly for several child transactions.

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-web3/src/test/java/com/hedera/services/store/contracts/precompile/utils/PrecompilePricingUtilsTest.java [hunk 1]
```diff
@@ -57,20 +55,20 @@ class PrecompilePricingUtilsTest {
...
     @Mock
-    private Provider<FeeCalculator> feeCalculator;
+    private FeeCalculator feeCalculator;
 
     @Mock
     private BasicFcfsUsagePrices resourceCosts;
 
     @Mock
-    private StackedStateFrames<?> state;
+    private StackedStateFrames<Object> state;
```
Comment (0xivanov) [inline hedera-mirror-web3/src/test/java/com/hedera/services/store/contracts/precompile/utils/PrecompilePricingUtilsTest.java:64]: Remove

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/MultiAssociatePrecompile.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        final var tokenIDs = decodeTokenIDsFromBytesArray(decodedArguments.get(1));
+
+        final var associateOp = Association.multiAssociation(accountID, tokenIDs);
+
+        return createAssociate(associateOp);
+    }
+
+    @Override
+    public long getGasRequirement(final long blockTimestamp, final Store store) {
+        //        return pricingUtils.computeGasRequirement(blockTimestamp, this);
```
Comment (jnels124) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/MultiAssociatePrecompile.java:65]: fix/remove

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServicePrecompileTest.java [hunk 3]
```diff
@@ -75,6 +76,21 @@ void evmPrecompileUnsupportedModificationTokenFunctionsTest(UnsupportedContractM
...
+    void evmPrecompileSupportedModificationTokenFunctionsTest(SupportedContractModificationFunctions contractFunc) {
+        final var functionHash = functionEncodeDecoder.functionHashFor(
+                contractFunc.name, MODIFICATION_CONTRACT_ABI_PATH, contractFunc.functionParameters);
+        final var serviceParameters = serviceParametersForEthEstimateGas(functionHash);
+
+        final var expectedGasUsed = gasUsedAfterExecution(serviceParameters);
+
+        assertThat(longValueOf.applyAsLong(contractCallService.processCall(serviceParameters)))
+                .as("result must be within 5-20% bigger than the gas used from the first call")
+                .isGreaterThanOrEqualTo((long) (expectedGasUsed * 1.05)) // expectedGasUsed value increased by 5%
```
Comment (jnels124) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServicePrecompileTest.java:90]: should this adjustment be configurable?

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+/**
+ * This class is a modified copy of HTSPrecompiledContract from hedera-services repo. Additionally,
+ * it implements an adapter interface which is used by {@link com.hedera.mirror.web3.evm.store.contract.precompile.MirrorHTSPrecompiledContract}.
+ * In this way once we start consuming libraries like smart-contract-service it would be easier to
+ */
+public class HTSPrecompiledContract implements HTSPrecompiledContractAdapter {
+
+    public static final PrecompileContractResult INVALID_DELEGATE = new PrecompileContractResult(
+            null, true, MessageFrame.State.COMPLETED_FAILED, Optional.of(ExceptionalHaltReason.PRECOMPILE_ERROR));
+    private static final Logger log = LogManager.getLogger(HTSPrecompiledContract.class);
```
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java:66]: Shouldn't use lombok inside copied services code

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/utils/contracts/ParsingConstants.java [hunk 5]
```diff
@@ -21,16 +21,17 @@
...
-
+    public static final String INT = "(int)";
     public static final TupleType burnReturnType = TupleType.parse("(int32,uint64)");
     public static final TupleType mintReturnType = TupleType.parse("(int32,uint64,int64[])");
     public static final TupleType hapiAllowanceOfType = TupleType.parse("(int32,uint256)");
     public static final TupleType intAddressTuple = TupleType.parse("(int32,address)");
     public static final TupleType intTuple = TupleType.parse(INT32);
 
+    private ParsingConstants() {
+        throw new UnsupportedOperationException("Utility class");
```
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-web3/src/main/java/com/hedera/services/hapi/utils/contracts/ParsingConstants.java:32]: Shouldn't use lombok inside copied services code.

**Review by xin-hedera — APPROVED** [review]

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        }
+        final var result = computePrecompile(input, frame);
+        return Pair.of(gasRequirement, result.getOutput());
+    }
+
+    @NonNull
+    public PrecompileContractResult computePrecompile(final Bytes input, @NonNull final MessageFrame frame) {
+        /* TODO Temporary workaround allowing eth_call to execute precompile methods in a dynamic context (non pure/view).
+        This is done by calling ViewExecutor/RedirectViewExecutor logic instead of Precompile classes.
+        After the Precompile classes are implemented, this workaround won't be needed. */
```
[the remark this answers is not in the document]
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java:129]: We have an [epic](https://github.com/hashgraph/hedera-mirror-node/issues/5504).

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/MultiAssociatePrecompile.java (hunk 2 above)
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/MultiAssociatePrecompile.java]: Removed

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServicePrecompileTest.java (hunk 3 above)
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServicePrecompileTest.java:90]: I think we have defined this offset between 5% and 20% as a best estimate ratio. Probably it's unlikely to change these values. If we need so, we can add configurable fields in a separate PR.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/config/EvmConfiguration.java [hunk 7]
```diff
@@ -133,18 +146,37 @@ PricedUsageCalculator pricedUsageCalculator(
         return new PricedUsageCalculator(accessorBasedUsages, overflowCheckingCalc);
     }
 
+    @Bean
+    EstimatorFactory estimatorFactory() {
```
[the remark this answers is not in the document]
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/config/EvmConfiguration.java:150]: I agree. We can create a new PR for this change.

File: docs/configuration.md [hunk 8]
```diff
@@ -581,6 +581,7 @@ value, it is recommended to only populate overridden properties in the custom `a
  `hedera.mirror.web3.evm.maxTokensPerAccount`           | 1000                                               | Maximum number token associations per account                                                                                                                                                 
  `hedera.mirror.web3.evm.network`                       | TESTNET                                            | Which Hedera network to use. Can be either `MAINNET`, `PREVIEWNET`, `TESTNET` or `OTHER`                                                                                                      
  `hedera.mirror.web3.evm.rateLimit`                     | 100s                                               | Maximum RPS limit                                                                                                                                                                             
+ `hedera.mirror.web3.evm.tokenAssociationsLimited`      | true                                               | Limits the number of token associations per account when true
```
[the remark this answers is not in the document]
Comment (IvanKavaldzhiev) [reply docs/configuration.md]: Removed. It was previously removed, probably it appeared again after a merge conflict.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/CachingStateFrame.java [hunk 9]
```diff
@@ -25,6 +25,7 @@
 import java.util.Optional;
 import java.util.stream.Collectors;
 import lombok.NonNull;
+import org.checkerframework.checker.units.qual.K;
```
[the remark this answers is not in the document]
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/CachingStateFrame.java]: Removed

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractAssociatePrecompile.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+import com.hedera.services.store.models.Id;
+import com.hedera.services.txn.token.AssociateLogic;
+import com.hedera.services.utils.EntityIdUtils;
+import com.hederahashgraph.api.proto.java.Timestamp;
+import com.hederahashgraph.api.proto.java.TokenAssociateTransactionBody;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import java.util.Objects;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+
+public abstract class AbstractAssociatePrecompile implements Precompile {
```
Comment (0xivanov) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractAssociatePrecompile.java:37]: Add `javadoc` and list the differences between this mirror-node vs services version of this class.

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractAssociatePrecompile.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        this.mirrorNodeEvmProperties = mirrorNodeEvmProperties;
+    }
+
+    @Override
+    public long getMinimumFeeInTinybars(Timestamp consensusTime) {
+        return pricingUtils.getMinimumPriceInTinybars(ASSOCIATE, consensusTime);
+    }
+
+    @Override
+    public RunResult run(MessageFrame frame, final Store store, final TransactionBody transactionBody) {
```
Comment (0xivanov) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractAssociatePrecompile.java:54]: In services we invoke `validateSyntax` method on the `transactionBody` to check it's validity. We might need that validation here as well.

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AssociatePrecompile.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.TokenID;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.UnaryOperator;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.datatypes.Address;
+
+public class AssociatePrecompile extends AbstractAssociatePrecompile {
```
Comment (0xivanov) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AssociatePrecompile.java:46]: Add `javadoc`

**Review by 0xivanov — COMMENTED** [review]
Looks good!

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AssociatePrecompile.java (hunk 12 above)
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AssociatePrecompile.java:55]: Added

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractAssociatePrecompile.java (hunk 11 above)
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractAssociatePrecompile.java:66]: Added

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractAssociatePrecompile.java (hunk 10 above)
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractAssociatePrecompile.java:49]: Added

**Review by 0xivanov — APPROVED** [review]
LGTM

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Pull request merged**